File: exercism/scala/secret-handshake/src/main/scala/secretHandshake.scala
import scala.util._

object SecretHandshake {
  val cmdMap = Map(
    1 -> "wink",
    2 -> "double blink",
    4 -> "close your eyes",
    8 -> "jump"
  )

  def handshake(n: String): List[String] =
    Try(handshake(Integer.parseInt(n, 2))).recover { case _ => Nil } get

  def handshake(n: Int): List[String] = n match {
    case 0 => Nil
    case n if n == 1 || n % 2 == 0 => List(cmdMap(n))
    case _ =>
      val m = n & 15 // split last for digit (0 | 0000)
      val ls = Stream.iterate((m + 1) >> 1)(_ >> 1).takeWhile(_ > 0).toList

      if (n > 15) ls.flatMap(handshake) else ls.flatMap(handshake).reverse
  }
}

File: exercism/haskell/secret-handshake/src/SecretHandshake.hs
module SecretHandshake (handshake) where

import Data.Bits
import Data.Maybe
import Data.Bool

class CanBit a where
  toBool :: a -> Maybe Bool

instance CanBit Char where
  toBool '0' = Just False
  toBool '1' = Just True
  toBool _ = Nothing

instance CanBit a => Handshake [a] where
  interpret = reverse . fromMaybe [] . traverse toBool

class Handshake a where
  interpret :: a -> [Bool]

instance Handshake Int where
  interpret n = map (testBit n) [0..numBits]
    where
      numBits = finiteBitSize n - countLeadingZeros n - 1

handshake :: Handshake a => a -> [String]
handshake = foldl (flip ($)) [] . zipWith (bool id) ops . interpret
  where
    ops = map snoc ["wink", "double blink", "close your eyes", "jump"] ++ [reverse]
    snoc s = (++ [s])
